### ISSUE TITLE
FR: Gerrit support

### ISSUE THREAD

**Issue by ilyagr** [martinvonz/jj#4387]
I'm creating a placeholder bug to reference. See also https://martinvonz.github.io/jj/latest/FAQ/#how-do-i-integrate-jujutsu-with-gerrit .

This would likely be a collection of `jj gerrit` commands.

Cc @thoughtpolice

**Comment by steveklabnik** [comment]
See also https://github.com/martinvonz/jj/pull/2845

**Comment by PhilipMetzger** [comment]
And also related #485.

**Comment by dbarnett** [comment]
What's the scope of Gerrit-specific functionality we're planning to integrate? From the links above it looks like `gerrit send` and/or something for rewriting commit messages on push, and then something for pushing stacked reviews.

I just filed a corresponding GitHub issue #4555 so I'm wondering what functionality would have Gerrit equivalents (and also whether `jj [forgename] foo` commands for those things are really necessary or should be modeled at another level).

**Comment by avamsi** [comment]
Mirroring https://github.com/martinvonz/jj/pull/2845#issuecomment-2543447845:

> In case it helps anyone, I set `Change-Id` using `templates.draft_commit_description` and just `git push`:
>
> ```toml
> [templates]
> 	draft_commit_description = '''
> 		separate("\n",
> 			description.remove_suffix("\n"),
> 			if(!description.contains(change_id.normal_hex()),
> 				"\nChange-Id: " ++ gerrit_change_id(change_id)
> 			),
> 			"\n",
> 			surround("JJ: Changes:\n", "", indent("JJ: \t", diff.summary()))
> 		)
> 	'''
> ```
> 
> (https://github.com/avamsi/dotfiles/blob/28a0e40439de56fe179a1b0e705727c85e075464/.jjconfig.toml#L146-L156)

**Comment by Benjamin-Davies** [comment]
Perhaps there should be a nice way to download the latest commit for a given change (referenced by either the Gerrit "Change-Id" or its numeric id).

The newly-fetched commits should be assigned a consistent Jujutsu change id. This could just be the reverse of however the "Change-Id" is generated on push.

**Comment by prattmic** [comment]
I've been using `jj` with Gerrit for over a year now and I think various improvements over the last year have made the Gerrit experience much smoother, even without an explicit Gerrit subcommand (#2845).

Notably:

* With the switch to using a git subprocess for fetch, authentication pain is gone. As long as git authenticates, jj does as well. I used to have two remotes: one unauthenticated I used for `jj git fetch`, another authenticated I used for `git push`.
* With the new `commit_trailers` and `format_gerrit_change_id_trailer` templates (#6088), it's now very easy to get `Change-Id` support (even more so than https://github.com/jj-vcs/jj/issues/4387#issuecomment-2543448225 which was honestly fine too).

For completeness, these are the things that I find are still pain points:

1. The obvious one is the inability to push to an arbitrary ref makes it impossible to actually push changes to Gerrit. But I hesitate to actually call this a "pain point" because I use a colocated repo and `git push origin HEAD:refs/for/master` and _it works perfectly fine_. I honestly don't have any issues with the way this works. It's primarily just "ugly" to need to use git directly when I do everything else through jj. Plus there is a minor baseline fear with a colocated repo that I will accidentally mutate git state in some way that breaks jj,

2. With regards to Change-Ids, I'm coming around to @thoughtpolice's position in https://github.com/jj-vcs/jj/pull/2845#issuecomment-2543359105 that having a Change-Id based directly on the jj change ID (as `format_gerrit_change_id_trailer` does) is a bad idea.

In addition to the abandoned and replaced case @thoughtpolice mentions, the place I run into this is splitting changes. Suppose I upload a change to Gerrit but later realize I want to split out a minor typo fix into a different change. I `jj split -i` the change, selecting all the main (non-typo-fix) changes for the first half of the split. I leave it's commit message and Change-Id alone. The second half of the split is the typo fix. I give it a completely new commit message and delete the old Change-Id. Finally I need one extra `jj describe` to re-run the `commit_trailers` template and replace the Change-Id I deleted.

This is all fine and works perfectly.

Now suppose I did this in the reverse order. In `jj split -i` I select the typo change first, replace the commit message and delete the Change-Id line. When `jj describe` reruns the template, it is going to add back the exact same Change-Id! That's because the first half of a split keeps its original jj change ID.

I think it would be better for the Change-Id to be random (I think based on the commit id would work fine) so that deleting and regenerating a Change-Id is a consistent way to get a different one. Of course, I can do this myself today by simply defining my own template instead of using `format_gerrit_change_id_trailer`.

(I acknowledge that there is an alternate viewpoint here that jj change IDs and Gerrit Change-Ids are conceptually very similar: a stable identifier for commits that change over time. If I want a set of changes conceptually associated with a specific Gerrit Change-Id then I should also want them conceptually associated with with the specific jj change ID. Thus I should simply not do the split in the "wrong" order. That's totally fair, but honestly just not something I'm thinking about closely most of the time. As a user, for my user experience it rarely matters that the change IDs are stable because I tend to look them up in the log right when I need them. The stability is of course important for internal operations, but for UX I wouldn't notice too much if they changed. Gerrit Change-Ids on the other hand are associated with other out-of-band data like code review comments and email subject lines that never update, so it is very noticeable if I were to move the main part of a change to a new Change-Id.)

3. Downloading changes is probably my one true pain point. I don't do this often, but it is painful enough that I simply don't do it in a jj repo and use a separate git clone instead. (I also haven't tried this in a while, so perhaps it has already improved)

The first line problem is that `jj git fetch` can't fetch from arbitrary refs, but like push I can use `git fetch` directly and that mostly works fine.

The real trouble comes when there are complicated dependencies.

Suppose I've uploaded patch set 1 of change A to Gerrit. My coworker downloads my change, adds another commit on top of it, which they upload to Gerrit as change B.

In the meantime, I make some adjustments to change A, which I upload to Gerrit as patch set 2.

Later, I decide I want to try out my coworker's change B, so I fetch it into my jj repo.

Now we have a problem. Change B depends on an _old version_ of change A. This makes jj really upset because this isn't normally possible in jj's model and makes references to change A ambiguous. To jj's credit, it never does anything wrong as far as I can tell, but any time change A is mentioned, jj asks which version of change A I meant. If I'm actually working on change A that gets really old and really annoying fast.

Admittedly, I'm not sure what jj should do here, the situation really is ambiguous. My initial thought is that references to change A should always refer to the latest version of A, and if old versions are reachable they should be given some modified change ID (e.g., `A~1`). But then that is adding a whole new layer of syntax for a situation that is normally impossible.

**Comment by martinvonz** [comment]
There has been *a lot* of discussion about many of these things lately.

One recent change is that we will start writing the jj change id to a `change-id` Git commit header (which means it's not part of the description like footers are). That's now opt-in (`git.write-change-id-header = true`) but it will be enabled by default from the next release. Gerrit will probably learn to read the change id from that header eventually. They have a design doc [here](https://gerrit-review.googlesource.com/c/homepage/+/464287).

Regarding which commit keeps the change id, there's open PR #6458 for adding `-A/-B/-d` flags to `jj split`. When given any of those flags, the selected changes will go into the new commit that is created in the specified location. The commit with the remaining changes will keep the change id. Once that's been merged, we can all try those flags and see how we like it. In particular, `jj split -B @` will behave similar to the current `jj split` but it will put the change id on the child commit.

Your last scenario (fetching an old version of commit from the remote) is interesting. One possible solution is to rebase your coworker's change B on top of your updated change A. I don't know if we would want to try to do that automatically. It could possibly be done by some new command designed to try to resolve divergence.

Btw, what does Gerrit do if your coworker rewrites change A and pushes change B based on that? Will Gerrit reject it since you own change A?

**Comment by ramenhost** [comment]
I recently replaced my git workflow with Jujutsu, so a `jj` newbie here. I would like support for fetching arbitrary Gerrit CLs. 

I have the below in mind (modelled after the commands in Gerrit's [download button](https://gerrit-review.googlesource.com/Documentation/images/user-review-ui-change-screen-download-commands.png))
- `jj gerrit fetch <change>[/<patchset>]` — Brings the changeset ref (refs/changes/NN/CCCC/PP) into the repo and associates it with a jj change-id, but does not move `@`.
- `jj gerrit checkout <change>[/<patchset>]` — fetch + `jj edit` on the result.
- `jj gerrit duplicate <change>[/<patchset>]` — fetch + `jj duplicate` on the result.

Another option is to just use a generic command like #4388 instead. I think both will suffer from #7587 though.

I can start with some PRs but wanted to get an initial opinion on this. Maybe this is already being worked upon as I could see a couple of PRs to `gerrit` subcommand by @matts1.

**Comment by matts1** [comment]
> * `jj gerrit fetch <change>[/<patchset>]` — Brings the changeset ref (refs/changes/NN/CCCC/PP) into the repo and associates it with a jj change-id, but does not move `@`.

Agreed, we'd definitely like this feature. If you want to implement this, happy to review it. Been a bit too busy to work on jj for the past month or two. We definitely want an actual gerrit command.

Given that the upload command is called `upload` though, I think we should probably call it `download` rather than fetch.

> * `jj gerrit checkout <change>[/<patchset>]` — fetch + `jj edit` on the result.

I agree that this is generally useful, but I don't like the use of the word "checkout". Two options come to mind here that I prefer, not sure which one is better.
* `jj gerrit new` / `jj gerrit edit` I think provides a lot more clarity, and is consistent with jj's naming scheme. It also distinguishes between new and edit workflows.
* `jj gerrit download --new / --edit` simplifies it into a single command, but I'm a little hesitant because it feels like something that might be outside of the scope of "download"

> * `jj gerrit duplicate <change>[/<patchset>]` — fetch + `jj duplicate` on the result.

Is what you actually want to duplicate it? Or to detach it from the existing commit so that re-uploading creates a new commit? The former seems too niche to be its own command, while the latter would probably be better achieved via `jj gerrit download --detach` IMO. I think that's well within the scope of the download command

If you want to work on this, just start with a simple `jj gerrit fetch` that downloads from gerrit without any flags.

**Comment by neild** [comment]
I spent some time poking around trying to figure out how I'd expect a `jj download` to behave, and I think there's a key question that I don't know the answer to:

Should all the revisions of a Gerrit CL map to the same jj change? Or is each revision a distinct change?

Right now (at least on the Gerrit instance I use) the answer depends on whether the revisions were uploaded by jj or by git.
- When jj uploads a new revision, it sets a `change-id` header, so all jj-uploaded revisions are the same change.
- When git uploads a new revision, it sets a `Change-Id:` trailer but no header. jj pays no attention to that trailer, so all git-uploaded revisions are distinct changes.

This is observed in practice if you download multiple revisions of a CL from Gerrit into a jj repo: If the revisions were created by jj they appear as divergent changes (same change-id, different content) and if they were created by git they appear as non-divergent changes with distinct jj change-ids.

What is the desired model? Should CL/1234/1 and CL/1234/2 be the same change? Or different ones? I think this should be consistent regardless of who uploaded those revisions.

In answering this, I think it's useful to draw a distinction between a *mutable* CL (CL/1234, a work in progress) and an *immutable* revision (CL/1234/1, a snapshot uploaded to Gerrit, can never change).

A simple model is: A mutable CL is a jj change and an immutable revision is a different jj change. You work on jj change A, you upload it as CL/1234/1, you download CL/1234/2 which is jj change B. If you want to update your work in progress from what you downloaded, you can run `jj restore --to A --from B`.

A more Gerrit-aware model might be: A CL is a jj change, and a change can contain immutable revisions (in its evolog?). You work on change A, you upload it as CL/1234/1, you download CL/1234/2 which is now in jj change A's evolog. Perhaps you can reference this as `A/2`, and update your WIP from the download with `jj restore --to A --from A/2`. I'm pretty sure this requires adding some new concept to jj to express the revisions of a change; I'm not enough of a jj expert to say how you'd go about this.

**Comment by FiloSottile** [comment]
I would strongly expect different Gerrit patch sets of the same CL to be fetched as commits associated with the same change ID.

As @neild said, this is already how it works if the patch sets were developed locally or were mailed with a `change-id` header. `jj gerrit fetch` would just need to adjust the internal representation for patch sets that lack the header.

**Comment by josephlou5** [comment]
> * When jj uploads a new revision, it sets a `change-id` header, so all jj-uploaded revisions are the same change.
> * When git uploads a new revision, it sets a `Change-Id:` trailer but no header. jj pays no attention to that trailer, so all git-uploaded revisions are distinct changes.

Btw, `jj gerrit upload` sets the `Change-Id:` trailer as well (it's what Gerrit uses to identify the same change), but it's not saved in your local commit description. See the last paragraph of https://docs.jj-vcs.dev/latest/gerrit/#change-id-management.

Also while I'm here I guess I'll say that I would expect a Gerrit CL to have the same jj change ID throughout its lifetime, but as you pointed out this can be hard if it was uploaded with Git (i.e., lacks jj's `change-id` header). In theory we could try to reverse-engineer a generated `Change-Id` from `jj gerrit upload`, but that's not necessarily guaranteed to be right (a randomly-generated `Change-Id` could still end in the same hex we pad with).

**Comment by neild** [comment]
> I would expect a Gerrit CL to have the same jj change ID throughout its lifetime

I want to emphasize again that there's a difference between a *CL* which changes over its lifetime, and a *revision* which is a specific snapshot of that CL.

CL/1234 changes. CL/1234/1 is created once and then can never again change.

My question is whether CL/1234, CL/1234/1, and CL/1234/2 should all share the same jj change or not.

**Comment by josephlou5** [comment]
> > I would expect a Gerrit CL to have the same jj change ID throughout its lifetime
> 
> I want to emphasize again that there's a difference between a *CL* which changes over its lifetime, and a *revision* which is a specific snapshot of that CL.
> 
> CL/1234 changes. CL/1234/1 is created once and then can never again change.
> 
> My question is whether CL/1234, CL/1234/1, and CL/1234/2 should all share the same jj change or not.

Yes, I'm aware. That's why I said "throughout its lifetime": each fetched patch set should (IMO) be a distinct _commit / revision_ of the same _change_. This ensures that downloading the same CL twice (at different patch sets, for example) properly updates the previously associated change for the CL. But if that's hard to do (or can't be determined), then I think it would be fine to have a new change ID.

**Comment by neild** [comment]
Apologies if that came across as condescending; I just want to understand precisely when someone is talking about a CL or a revision, because it's not always clear and it matters.

> downloading the same CL twice (at different patch sets, for example) properly updates the previously associated change for the CL

Do you mean that if I have jj change A associated with CL 1, then downloading revision CL/1/1 should modify the current contents of A? Or that it should leave the current contents of A alone, but associate the contents of CL/1/1 with A? If the latter, should CL/1/1 go into A's evolog or somewhere else?

I do think that downloading a revision and updating the change associated with that revision's CL need to be distinct operations. You might want to download a prior revision of a change to, for example, diff it against your current work. Or you might want to download several revisions of a change to diff against each other. Or the CL might be submitted, in which case the submitted revision is an immutable component of history and downloading a different revision definitely can't modify it.

**Comment by josephlou5** [comment]
> Apologies if that came across as condescending; I just want to understand precisely when someone is talking about a CL or a revision, because it's not always clear and it matters.

No worries! We are dealing with two different systems here with overlapping terms but differing definitions, so it is bound to be confusing.

I have created a new child issue #10116 for `jj gerrit download`, so discussion can move there and keep this parent issue clean. I will reply to your comment there.
